### PR TITLE
lib: Convert cockpit-components-inline-notification to TypeScript and functional component

### PR DESCRIPTION
**Pull request by martinpitt**
Plus two other small fixes. Broken out of #20726 to bikeshed about the typing separately.

### CODE REVIEW

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 1]
```diff
@@ -26,17 +26,30 @@ import './cockpit-components-inline-notification.css';
 
 const _ = cockpit.gettext;
 
-function mouseClick(fun) {
-    return function (event) {
+function mouseClick(fun: (ev: React.MouseEvent) => void): (ev: React.MouseEvent) => void {
+    return function (event: React.MouseEvent) {
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:30]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 2]
```diff
@@ -26,17 +26,30 @@ import './cockpit-components-inline-notification.css';
...
+    isInline?: boolean;
+    isLiveRegion?: boolean;
+}
+
+interface InlineNotificationState {
+    isDetail: boolean;
+}
+
+export class InlineNotification extends React.Component<InlineNotificationProps, InlineNotificationState> {
+    constructor(props: InlineNotificationProps) {
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:52]: This added line is not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 3]
```diff
@@ -66,11 +79,12 @@ export class InlineNotification extends React.Component {
         }
         const extraProps = {};
         if (onDismiss)
+            // @ts-expect-error: PF bug: actionClose not defined in Alert
             extraProps.actionClose = <AlertActionCloseButton onClose={onDismiss} />;
 
         return (
             <Alert variant={type || 'danger'}
-                isLiveRegion={this.props.isLiveRegion}
+                isLiveRegion={this.props.isLiveRegion ?? false}
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:87]: This added line is not executed by any test.

Comment (martinpitt) [reply pkg/lib/cockpit-components-inline-notification.tsx:87]: They are covered in #20726.

File: pkg/lib/cockpit-components-inline-notification.tsx (hunk 1 above)
Comment (allisonkarlitskaya) [reply pkg/lib/cockpit-components-inline-notification.tsx:30]: I just don't like this function.  This whole setup with mouseClick() and toggleDetail() and `this.toggleDetail = this.toggleDetail.bind(this)` is way too complicated.  Now is a good opportunity to clean this up.

My instinct is always towards inlining handlers with typescript to benefit from the inferred types for the parameters.  I tried a bit at that and was able to remove all of the complicated tangle.  I don't know if this works, but I managed to get to something like:

```tsx
        const detailButton = (detail &&
            <Button variant="link" isInline className='alert-link more-button'
                onClick={event => {
                    if (event.button !== 0)
                        return;
                    event.preventDefault();
                    this.setState({ isDetail: !this.state.isDetail });
                }}
            >
                {this.state.isDetail ? _("show less") : _("show more")}
            </Button>
        );
```

...at this point, the `detailButton` itself could probably be inlined into the main component, but that's a matter of taste...

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 4]
```diff
@@ -26,17 +26,30 @@ import './cockpit-components-inline-notification.css';
...
         return fun(event);
     };
 }
 
-export class InlineNotification extends React.Component {
-    constructor(props) {
+interface InlineNotificationProps {
+    text: string;
+    detail?: string;
+    type?: "danger" | "success" | "warning" | "info" | "custom";
```
Comment (allisonkarlitskaya) [inline pkg/lib/cockpit-components-inline-notification.tsx:41]: I feel like there is some way to hit this with `typeof AlertProps['variant']` or so.  @jelly made a similar request to me in another PR recently...

**Comment by martinpitt** [comment]
> Does it unblock you, or is it just a chance to get a few extras landed in order to decrease the size of the other work?

Just the latter, to get them out of they way, and they seem mildly useful/correct on their own.

For some reason I can't respond to your "mouseClick" thread (github bug?) but I'm ok with inlining this.

File: pkg/lib/cockpit-components-inline-notification.tsx (hunk 4 above)
Comment (martinpitt) [reply pkg/lib/cockpit-components-inline-notification.tsx:41]: That works, thanks!

**Comment by martinpitt** [comment]
> to convert this to a functional component

Running out of time, and frankly also interest -- this feels like busywork to me at least right now, I really want to finish #20726. But I did the other two cleanups you suggested, thanks!

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 5]
```diff
@@ -20,57 +20,57 @@ import React from 'react';
...
 }
 
-export class InlineNotification extends React.Component {
-    constructor(props) {
+interface InlineNotificationState {
+    isDetail: boolean;
+}
+
+export class InlineNotification extends React.Component<InlineNotificationProps, InlineNotificationState> {
+    constructor(props: InlineNotificationProps) {
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:43]: This added line is not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 6]
```diff
@@ -20,57 +20,57 @@ import React from 'react';
...
-            if (this.state.isDetail) {
-                detailButtonText = _("show less");
-            }
+        const detailButton = (detail &&
+            <Button variant="link" isInline className="alert-link more-button"
+                onClick={event => {
+                    if (event.button !== 0)
+                        return;
+                    event.preventDefault();
+                    this.setState(old => { return { isDetail: !old.isDetail } });
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:59]: These 7 added lines are not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 7]
```diff
@@ -20,57 +20,57 @@ import React from 'react';
...
+                }}
+            >
+                {this.state.isDetail ? _("show less") : _("show more")}
+            </Button>
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:63]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 8]
```diff
@@ -20,57 +20,57 @@ import React from 'react';
...
-        }
         const extraProps = {};
         if (onDismiss)
+            // @ts-expect-error: PF bug: actionClose not defined in Alert
             extraProps.actionClose = <AlertActionCloseButton onClose={onDismiss} />;
 
         return (
             <Alert variant={type || 'danger'}
-                isLiveRegion={this.props.isLiveRegion}
+                isLiveRegion={this.props.isLiveRegion ?? false}
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:73]: This added line is not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx (hunk 5 above)
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:43]: This added line is not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx (hunk 6 above)
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:59]: These 7 added lines are not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx (hunk 7 above)
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:63]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx (hunk 8 above)
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:73]: This added line is not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 9]
```diff
@@ -20,57 +20,57 @@ import React from 'react';
 import PropTypes from 'prop-types';
 import cockpit from 'cockpit';
 
-import { Alert, AlertActionCloseButton } from "@patternfly/react-core/dist/esm/components/Alert/index.js";
+import { Alert, AlertActionCloseButton, AlertProps } from "@patternfly/react-core/dist/esm/components/Alert/index.js";
```
Comment (jelly) [inline pkg/lib/cockpit-components-inline-notification.tsx:23]: This is more TIL a few days ago, you can import this as:

`import type AlertProps` so it won't end up in bundled JS.

Comment (martinpitt) [reply pkg/lib/cockpit-components-inline-notification.tsx]: Done, thanks! I fixed the other instance in pkg/lib/cockpit-components-dropdown.tsx as well.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+import cockpit from 'cockpit';
+
+import { Alert, AlertActionCloseButton } from "@patternfly/react-core/dist/esm/components/Alert/index.js";
+import type { AlertProps } from "@patternfly/react-core/dist/esm/components/Alert/index.js";
+import { Button } from "@patternfly/react-core/dist/esm/components/Button/index.js";
+import './cockpit-components-inline-notification.css';
+
+const _ = cockpit.gettext;
+
+export const InlineNotification = ({ text, detail, type = "danger", onDismiss, isInline = true, isLiveRegion = false }: {
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:30]: This added line is not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    text: string;
+    detail?: string;
+    type: AlertProps["variant"];
+    onDismiss?: (ev?: Event) => void;
+    isInline: boolean;
+    isLiveRegion: boolean;
+}) => {
+    const [isDetail, setIsDetail] = useState(false);
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:38]: This added line is not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+    const detailButton = (detail &&
+        <Button variant="link" isInline className="alert-link more-button"
+            onClick={event => {
+                if (event.button !== 0)
+                    return;
+                event.preventDefault();
+                setIsDetail(prev => !prev);
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:46]: These 7 added lines are not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+            }}
+        >
+            {isDetail ? _("show less") : _("show more")}
+        </Button>
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:50]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    );
+
+    const extraProps = {};
+    if (onDismiss)
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:54]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+        // @ts-expect-error: PF bug: actionClose not defined in Alert
+        extraProps.actionClose = <AlertActionCloseButton onClose={onDismiss} />;
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:56]: This added line is not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+    return (
+        <Alert variant={type || 'danger'}
+            isLiveRegion={isLiveRegion}
+            isInline={isInline}
+            title={<> {text} {detailButton} </>} {...extraProps}>
+            {isDetail && (<p>{detail}</p>)}
+        </Alert>
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:64]: These 7 added lines are not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+export const ModalError = ({ dialogError, dialogErrorDetail, id, isExpandable }: {
+    dialogError: string,
+    dialogErrorDetail?: string,
+    id?: string,
+    isExpandable?: boolean,
+}) => {
+    // HACK: id! here is a workaround, it's inconvenient to conditionally add it
+    return (
+        <Alert id={id!} variant='danger' isInline title={dialogError} isExpandable={!!isExpandable}>
+            { typeof dialogErrorDetail === 'string' ? <p>{dialogErrorDetail}</p> : dialogErrorDetail }
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:86]: This added line is not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx (hunk 15 above)
Comment (allisonkarlitskaya) [reply pkg/lib/cockpit-components-inline-notification.tsx:56]: Can you not write this inline like:

```jsx
<Alert
  etc
  ...{ onDismiss && <AlertActionClodeButton onClose={onDismiss} /> }
  etc
  >
```

?

Does that help with the types at all?

Comment (martinpitt) [reply pkg/lib/cockpit-components-inline-notification.tsx]: Yes, it does. Done.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 18]
```diff
@@ -66,11 +80,12 @@ export class InlineNotification extends React.Component {
         }
         const extraProps = {};
         if (onDismiss)
+            // @ts-expect-error: PF bug: actionClose not defined in Alert
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply pkg/lib/cockpit-components-inline-notification.tsx]: This does not exist any more, see above.

File: pkg/lib/cockpit-components-dropdown.tsx [hunk 19]
```diff
@@ -21,7 +21,8 @@ import React, { useState } from 'react';
 import PropTypes from "prop-types";
 
 import { MenuToggle } from "@patternfly/react-core/dist/esm/components/MenuToggle";
-import { Dropdown, DropdownList, DropdownPopperProps } from "@patternfly/react-core/dist/esm/components/Dropdown";
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply pkg/lib/cockpit-components-dropdown.tsx:24]: OK, I changed it back.

[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply pkg/lib/cockpit-components-dropdown.tsx:24]: For the record, I tested this in development and production builds.  In development output it made no difference at all.  In production, it caused the generated minified variable names to shift (but not change in size).  The change in the variable names caused the compressed output to change in size by one byte, though — but clearly that's luck.

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
The only real issue here is the `!`.  I don't think it's necessary, and we should avoid `!` and `as` unless it's absolutely strictly necessary.

We might open bugs in PatternFly to request that they consistently accept `undefined` (and maybe also `null`) as a value for optional values on their components... it would make this sort of thing a lot easier on our side.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+import PropTypes from 'prop-types';
+import cockpit from 'cockpit';
+
+import { Alert, AlertActionCloseButton, AlertProps } from "@patternfly/react-core/dist/esm/components/Alert/index.js";
+import { Button } from "@patternfly/react-core/dist/esm/components/Button/index.js";
+import './cockpit-components-inline-notification.css';
+
+const _ = cockpit.gettext;
+
+export const InlineNotification = ({ text, detail, type = "danger", onDismiss, isInline = true, isLiveRegion = false }: {
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:29]: This added line is not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+    text: string;
+    detail?: string;
+    type: AlertProps["variant"];
+    onDismiss?: (ev?: Event) => void;
+    isInline: boolean;
+    isLiveRegion: boolean;
+}) => {
+    const [isDetail, setIsDetail] = useState(false);
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:37]: This added line is not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+
+    const detailButton = (detail &&
+        <Button variant="link" isInline className="alert-link more-button"
+            onClick={event => {
+                if (event.button !== 0)
+                    return;
+                event.preventDefault();
+                setIsDetail(prev => !prev);
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:45]: These 7 added lines are not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+            }}
+        >
+            {isDetail ? _("show less") : _("show more")}
+        </Button>
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:49]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+    );
+
+    return (
+        <Alert variant={type || 'danger'}
+            isLiveRegion={isLiveRegion}
+            isInline={isInline}
+            title={<> {text} {detailButton} </>}>
+            alertActionClose={onDismiss && <AlertActionCloseButton onClose={onDismiss} />}
+            {isDetail && (<p>{detail}</p>)}
+        </Alert>
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:59]: These 8 added lines are not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+export const ModalError = ({ dialogError, dialogErrorDetail, id, isExpandable }: {
+    dialogError: string,
+    dialogErrorDetail?: string,
+    id?: string,
+    isExpandable?: boolean,
+}) => {
+    // HACK: id! here is a workaround, it's inconvenient to conditionally add it
+    return (
+        <Alert id={id!} variant='danger' isInline title={dialogError} isExpandable={!!isExpandable}>
+            { typeof dialogErrorDetail === 'string' ? <p>{dialogErrorDetail}</p> : dialogErrorDetail }
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:81]: This added line is not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 26]
```diff
@@ -79,17 +94,25 @@ export class InlineNotification extends React.Component {
...
+export const ModalError = ({ dialogError, dialogErrorDetail, id, isExpandable }: {
+    dialogError: string,
+    dialogErrorDetail?: string,
+    id?: string,
+    isExpandable?: boolean,
+}) => {
+    // HACK: id! here is a workaround, it's inconvenient to conditionally add it
     return (
-        <Alert id={id} variant='danger' isInline title={dialogError} isExpandable={!!isExpandable}>
+        <Alert id={id!} variant='danger' isInline title={dialogError} isExpandable={!!isExpandable}>
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply pkg/lib/cockpit-components-inline-notification.tsx]: Done.

File: pkg/lib/cockpit-components-dropdown.tsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply pkg/lib/cockpit-components-dropdown.tsx]: AH right, done.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 27]
```diff
@@ -21,22 +21,36 @@ import PropTypes from 'prop-types';
 import cockpit from 'cockpit';
 
 import { Alert, AlertActionCloseButton } from "@patternfly/react-core/dist/esm/components/Alert/index.js";
+import type { AlertProps } from "@patternfly/react-core/dist/esm/components/Alert/index.js";
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply pkg/lib/cockpit-components-inline-notification.tsx]: Dropped the other commit.

File: pkg/lib/cockpit-components-inline-notification.tsx (hunk 20 above)
Comment (martinpitt) [reply pkg/lib/cockpit-components-inline-notification.tsx:29]: Yes, the whole component is unused in Cockpit. It's used in files and machines, I locally tested it in the playground, see https://github.com/cockpit-project/cockpit/pull/20736#issuecomment-2222366801

**Review by allisonkarlitskaya — APPROVED** [review]
-13 lines.  Not a bad cleanup!

**Comment by martinpitt** [comment]
Minus 13 lines, but plus lots of types! :grin:

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 28]
```diff
@@ -0,0 +1,83 @@
...
+import PropTypes from 'prop-types';
+import cockpit from 'cockpit';
+
+import { Alert, AlertActionCloseButton, AlertProps } from "@patternfly/react-core/dist/esm/components/Alert/index.js";
+import { Button } from "@patternfly/react-core/dist/esm/components/Button/index.js";
+import './cockpit-components-inline-notification.css';
+
+const _ = cockpit.gettext;
+
+export const InlineNotification = ({ text, detail, type = "danger", onDismiss, isInline = true, isLiveRegion = false }: {
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:29]: This added line is not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 29]
```diff
@@ -0,0 +1,83 @@
...
+    text: string;
+    detail?: string;
+    type: AlertProps["variant"];
+    onDismiss?: (ev?: Event) => void;
+    isInline: boolean;
+    isLiveRegion: boolean;
+}) => {
+    const [isDetail, setIsDetail] = useState(false);
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:37]: This added line is not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 30]
```diff
@@ -0,0 +1,83 @@
...
+
+    const detailButton = (detail &&
+        <Button variant="link" isInline className="alert-link more-button"
+            onClick={event => {
+                if (event.button !== 0)
+                    return;
+                event.preventDefault();
+                setIsDetail(prev => !prev);
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:45]: These 7 added lines are not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 31]
```diff
@@ -0,0 +1,83 @@
...
+            }}
+        >
+            {isDetail ? _("show less") : _("show more")}
+        </Button>
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:49]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 32]
```diff
@@ -0,0 +1,83 @@
...
+    );
+
+    return (
+        <Alert variant={type || 'danger'}
+            isLiveRegion={isLiveRegion}
+            isInline={isInline}
+            title={<> {text} {detailButton} </>}
+            { ...onDismiss && { actionClose: <AlertActionCloseButton onClose={onDismiss} /> } }>
+            {isDetail && (<p>{detail}</p>)}
+        </Alert>
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:59]: These 8 added lines are not executed by any test.

File: pkg/lib/cockpit-components-inline-notification.tsx [hunk 33]
```diff
@@ -0,0 +1,83 @@
...
+
+export const ModalError = ({ dialogError, dialogErrorDetail, id, isExpandable }: {
+    dialogError: string,
+    dialogErrorDetail?: string,
+    id?: string,
+    isExpandable?: boolean,
+}) => {
+    return (
+        <Alert {...id && { id }} variant='danger' isInline title={dialogError} isExpandable={!!isExpandable}>
+            { typeof dialogErrorDetail === 'string' ? <p>{dialogErrorDetail}</p> : dialogErrorDetail }
```
Comment (cockpituous) [inline pkg/lib/cockpit-components-inline-notification.tsx:80]: This added line is not executed by any test.

**Pull request merged**